_posts/2019-09-03-how-to-deploy-production-grade-kubernetes-cluster-aws.adoc

```diff
@@ -0,0 +1,1850 @@
+  EKS bootstrap script so that those tags show up as labels in Kubernetes.
+
+https://github.com/gruntwork-io/module-aws-monitoring/tree/master/modules/metrics/cloudwatch-memory-disk-metrics-scripts[cloudwatch-memory-disk-metrics-scripts]::
+  Send memory and disk usage metrics for your EC2 Instances to CloudWatch. These metrics are not available by default
+  as they are only visible from inside a VM.
+
+https://github.com/gruntwork-io/module-aws-monitoring/tree/master/modules/logs/syslog[syslog]::
+  Configure log rotation and rate limiting for syslog.
+
+https://github.com/gruntwork-io/module-security/tree/master/modules/auto-update[auto-update]::
+  Configure a Linux server to automatically install critical security updates on a nightly basis.
+
+https://github.com/gruntwork-io/module-security/tree/master/modules/fail2ban[fail2ban]::
+  Configure a Linux server to automatically ban malicious ip addresses from connecting to the server via SSH.
+
+https://github.com/gruntwork-io/module-security/tree/master/modules/ntp[ntp]::
+  Install and configure NTP on a Linux server to prevent clock drift.
+
+https://github.com/gruntwork-io/module-security/tree/master/modules/ip-lockdown[ip-lockdown]::
+  Lock down specified IP addresses so only certain OS users can access them. Primarily used to lock down the EC2
+  instance metadata endpoint (and therefore, the IAM role attached to the EC2 instance) so that it can only be accessed
+  by specific users (e.g., only `root`).
+
+https://github.com/gruntwork-io/module-security/tree/master/modules/ssh-grunt[ssh-grunt]::
+  Allow managing SSH access to EC2 instances using IAM. Developers you add to specific IAM groups will be able to SSH
+  to specific servers using their own username and SSH key.
+
+To build an AMI from this Packer template, you run:
+
+----
+packer build eks-node.json
+----
+
+Packer will output the ID of the AMI at the end of the build. Copy this AMI down so you can deploy it later in this
+guide.
+
+=== Configure the worker node User Data script
+
+Now that you know what will be installed on each worker node AMI, you can fill in the User Data script that each worker
+node will run on boot. Create `user-data.sh` in your `infrastructure-modules` repo:
+
+----
+infrastructure-modules
+  └ networking
+    └ vpc-mgmt
+    └ vpc-app
+  └ services
+    └ eks-cluster
+      └ packer
+        └ eks-node.json
+      └ user-data
+        └ user-data.sh
+      └ main.tf
+      └ dependencies.tf
+      └ outputs.tf
+      └ variables.tf
+----
+
+Here's what `user-data.sh` should look like:
+
+.infrastructure-modules/services/eks-cluster/user-data/user-data.sh
+[source,bash]
+----
+#!/bin/bash
+
+set -e
+
+# Send the log output from this script to user-data.log, syslog, and the console
+# From: https://alestic.com/2010/12/ec2-user-data-output/
+exec > >(tee /var/log/user-data.log|logger -t user
```

Comment:
Roger! Let me know when you do!

Reply:
I just tried it out and indeed it failed, but unexpectedly. The VPC CNI plugin was what failed; everything else worked.

In my testing though I observed some weird/interesting behavior. The VPC CNI container was not able to access the metadata endpoint, which I wouldn't expect it to since it is a docker container. However, when I spun up a docker container and went in as root (e.g an alpine container, which only has the root user with the shell), I was able to access the metadata endpoint and get the instance AWS credentials. Is that expected?

Reply:
Oh, crap, missed this one. 

So you were able to run a Docker container on an EKS worker node, and the container was able to access EC2 metadata of the worker node? Even though you had run `ip-lockdown` or without?

Reply:
Yes even though I had run `ip-lockdown`. I know `ip-lockdown` was doing something because I noticed this when I booted the container after seeing a bunch of other containers report issues with the VPC CNI plugin, which went away when I removed the call in the `user-data`.

Reply:
Is the container being run by root or ec2-user or someone else with metadata access? Are you testing by running `curl http://169.254.169.254/latest/meta-data/`? Is it possible that request is being proxied by kube-proxy or something?

Reply:
There's also: https://docs.aws.amazon.com/eks/latest/userguide/restrict-ec2-credential-access.html

Reply:
> It looks like what is happening is that the user is resolving based on the user of the container. E.g a root user in the container looks the same as root on the instance inside iptables. So if you have a container that is using root as the exec user (E.g alpine linux containers) it has metadata access while containers that use a different user (most application containers) will not.

Why does the container have access to EC2 metadata on the host at all? Does EKS explicitly expose the host network to the container? If so, we should probably disable that as per the AWS docs. If you don't have it already, could you add a ticket to track that?

Reply:
> Does EKS explicitly expose the host network to the container?

The host network isn't exposed (unless you add the `host: true` spec to the pod). I don't exactly know how the instance profile is mapped to the pod ENI, but per the AWS docs:

> By default, containers that are running on your worker nodes are not prevented from accessing the credentials that are supplied to the worker node's instance profile through the Amazon EC2 instance metadata server.

Reply:
This by the way, is necessary for any application that doesn't support IRSA yet. Note that the aws-alb-ingress-controller hasn't updated the go SDK to support it.

Reply:
Intriguing. OK, so, how should we update this guide for now? Remove the `ip-lockdown` bit until we have time to implement it properly for EKS?

Reply:
I think we can keep it, although the purpose is less to prevent the containers from accessing the metadata, and more to prevent SSH users. The container prevention will be an amendment once we figure it all out. This might be a new script or new args to `ip-lockdown` that supports blocking network interfaces.

Reply:
Roger